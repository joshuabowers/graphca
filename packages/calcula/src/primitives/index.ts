import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
