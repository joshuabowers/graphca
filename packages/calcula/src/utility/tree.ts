import { Writer } from "../monads/writer"

export enum Clades {
  primitive = 'Primitive',  // Constants: 10, 4 - 5i, true
  unary = 'Unary',          // Functions and Ops: not, abs, sin
  binary = 'Binary',        // Functions and Ops: and, log, polygamma
  variadic = 'Variadic'     // Variables: x, OhGodWhy
}

export enum Genera {
  arithmetic = 'Arithmetic',
  connective = 'Connective',
  inequalities = 'Inequalities',
  trigonometric = 'Trigonometric',
  arcus = 'Arcus',
  hyperbolic = 'Hyperbolic',
  areaHyperbolic = 'AreaHyperbolic',
  combinatorics = 'Combinatorics'
}

export enum Species {
  // Primitives
  real = 'Real',
  complex = 'Complex',
  boolean = 'Boolean',
  nil = 'Nil',
  nan = 'NaN',

  // Arithmetic
  add = 'Addition',
  multiply = 'Multiplication',
  raise = 'Exponentiation',

  // Functions
  abs = 'Absolute',

  // :: Inequalities
  equals = 'Equality',
  notEquals = 'Inequality',
  lessThan = 'LessThan',
  greaterThan = 'GreaterThan',
  lessThanEquals = 'LessThanEquals',
  greaterThanEquals = 'GreaterThanEquals',

  // :: Connectives
  not = 'Complement',
  and = 'Conjunction',
  or = 'Disjunction',
  xor = 'ExclusiveDisjunction',
  implies = 'Implication',
  nand = 'AlternativeDenial',
  nor = 'JointDenial',
  xnor = 'Biconditional',
  converse = 'ConverseImplication',

  // :: Trigonometric
  sin = 'Sine',
  cos = 'Cosine',
  tan = 'Tangent',
  csc = 'Cosecant',
  sec = 'Secant',
  cot = 'Cotangent',

  // Variables
  variable = 'Variable'
}

export type TreeNode = {
  readonly clade: Clades,
  readonly genus?: Genera,
  readonly species: Species
}

export type DerivedNode<T extends ((...args: any[]) => any)> = 
  ReturnType<T> extends Writer<infer U> ? U : never

export type TreeNodeGuardFn<T extends TreeNode> = 
  (value: Writer<TreeNode>) => value is Writer<T>

// values should be Writer<TreeNode>, narrowed to Writer<T>
export const isClade = <T extends TreeNode>(clade: Clades) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.clade === clade

export const isGenus = <T extends TreeNode>(genus: Genera) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.genus === genus

export const isSpecies = <T extends TreeNode>(species: Species) =>
  (value: Writer<TreeNode>): value is Writer<T> =>
    value?.value.species === species

export const notAny = (...args: Species[]) => {
  const exclude = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && !exclude.has(t.value.species)
}

export const any = (...args: Species[]) => {
  const include = new Set(args)
  return <T extends TreeNode>(t: Writer<T>) => 
    t.value && include.has(t.value.species)
}
