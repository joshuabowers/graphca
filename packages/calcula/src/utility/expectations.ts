import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
