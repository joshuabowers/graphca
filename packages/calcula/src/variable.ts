import { Writer, unit } from './monads/writer'
import { TreeNode, Clades, Species, isSpecies } from './utility/tree'
import { isNil, nil } from './primitives/nil'
import { rule, process, resolve } from './utility/rule'

export type Variable = TreeNode & {
  readonly clade: Clades.variadic,
  readonly species: Species.variable,
  readonly name: string,
  readonly value: Writer<TreeNode>
}

export const isVariable = isSpecies<Variable>(Species.variable)

export const variable = (
  name: string, value: Writer<TreeNode> = nil
): Writer<Variable> => 
  ({
    value: {
      clade: Clades.variadic,
      species: Species.variable,
      name, value
    },
    log: [{
      input: process`${name}`, 
      rewrite: isNil(value) ? resolve`${name}` : process`${value}`,
      action: isNil(value) 
        ? 'referenced variable' : 'referenced variable with value'
    }] 
  })

export type Scope = Map<string, Writer<Variable>>

export const isScope = (v: unknown): v is Scope =>
  typeof v === 'object' && v !== null
  && ['get', 'set', 'has', 'delete'].every(m => m in v)

type Entries = Iterable<readonly [string, Writer<Variable>]>

export const scope = (entries: Entries = []): Scope => 
  new Map<string, Writer<Variable>>(entries)

export const assign = (name: string, value: Writer<TreeNode>, scope?: Scope) => {
  const v = variable(name, value)
  if(scope){
    if(isNil(value)){ scope.delete(name) }
    else { scope.set(name, v) }
  }
  return v
}
