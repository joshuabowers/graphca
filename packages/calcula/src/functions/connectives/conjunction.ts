import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when, binaryInfixRule } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { isComplement } from './complement'
import { isDisjunction } from './disjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Conjunction = Connective<Species.and>

export const andRule = binaryInfixRule(Unicode.and)

export const [and, isConjunction, $and] = binary<Conjunction, Boolean>(
  Species.and, Genera.connective
)(
  (l, r) => [boolean(l.value !== 0 && r.value !== 0), andRule(l, r), 'real conjunction'],
  (l, r) => [
    boolean((l.a !== 0 || l.b !== 0) && (r.a !== 0 || r.b !== 0)), 
    andRule(l, r),
    'complex conjunction'
  ],
  (l, r) => [boolean(l.value && r.value), andRule(l, r), 'boolean conjunction']
)(
  when(
    [_, isValue(boolean(true))], 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive identity']
  ),
  when(
    [isValue(boolean(true)), _], 
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive identity']
  ),
  when(
    [_, isValue(boolean(false))], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    [isValue(boolean(false)), _], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive idempotency']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  )
)
