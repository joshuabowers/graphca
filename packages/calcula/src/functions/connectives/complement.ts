import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation, isSpecies } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { Unary, unary, unaryFnRule, when } from "../../closures/unary"
import { Conjunction, and, andRule } from './conjunction'
import { Disjunction, or, orRule } from './disjunction'
import { ExclusiveDisjunction, xor, xorRule } from './exclusiveDisjunction'
import { Implication } from './implication'
import { AlternativeDenial, nand, nandRule } from './alternativeDenial'
import { JointDenial, nor, norRule } from './jointDenial'
import { Biconditional, xnor, xnorRule } from './biconditional'
import { ConverseImplication } from './converseImplication'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Complement = Unary<Species.not, Genera.connective>

export const notRule = unaryFnRule(Unicode.not)

export const [not, isComplement, $not] = unary<Complement, Boolean>(
  Unicode.not, Notation.prefix, Species.not, Genera.connective
)(
  r => [boolean(r.value === 0), notRule(r), 'real complement'],
  c => [boolean(c.a === 0 && c.b === 0), notRule(c), 'complex complement'],
  b => [boolean(!b.value), notRule(b), 'boolean complement']
)(
  // NOTE: Cannot use derived isSpecies guards as they are not defined
  // by this point.
  when<Complement>(
    isSpecies(Species.not), 
    v => [v.expression, rule`${v.expression}`, 'double complement']
  ),
  when<Conjunction>(
    isSpecies(Species.and), 
    v => [nand(v.left, v.right), nandRule(v.left, v.right), 'complement of conjunction']
  ),
  when<Disjunction>(
    isSpecies(Species.or), 
    v => [nor(v.left, v.right), norRule(v.left, v.right), 'complement of disjunction']
  ),
  when<AlternativeDenial>(
    isSpecies(Species.nand), 
    v => [and(v.left, v.right), andRule(v.left, v.right), 'complement of alternative denial']
  ),
  when<JointDenial>(
    isSpecies(Species.nor), 
    v => [or(v.left, v.right), orRule(v.left, v.right), 'complement of joint denial']
  ),
  when<ExclusiveDisjunction>(
    isSpecies(Species.xor), 
    v => [xnor(v.left, v.right), xnorRule(v.left, v.right), 'complement of exclusive disjunction']
  ),
  when<Implication>(
    isSpecies(Species.implies),
    v => [
      and(v.left, not(v.right)), 
      rule`${v.left} ${Unicode.and} (${v.right})${Unicode.not}`, 
      'complement of implication'
    ]
  ),
  when<Biconditional>(
    isSpecies(Species.xnor),
    v => [xor(v.left, v.right), xorRule(v.left, v.right), 'complement of biconditional']
  ),
  when<ConverseImplication>(
    isSpecies(Species.converse),
    v => [
      and(not(v.left), v.right), 
      rule`(${v.left})${Unicode.not} ${Unicode.and} ${v.right}`,
      'complement of converse implication'
    ]
  )
)
