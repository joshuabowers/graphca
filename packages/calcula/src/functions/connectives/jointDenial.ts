import { _ } from '@arrows/multimethod'
import { unit } from "../../monads/writer"
import { Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when, binaryInfixRule } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement, not, notRule } from './complement'
import { or } from './disjunction'
import { and, andRule } from './conjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type JointDenial = Connective<Species.nor>

export const norRule = binaryInfixRule(Unicode.nor)

export const [nor, isJointDenial, $nor] = binary<JointDenial, Boolean>(
  Unicode.nor, Notation.infix, Species.nor, Genera.connective
)(
  (l, r) => not(or(unit(l), unit(r))),
  (l, r) => not(or(unit(l), unit(r))),
  (l, r) => not(or(unit(l), unit(r)))
)(
  when(
    [_, isValue(boolean(true))],
    [boolean(false), rule`${boolean(false)}`, 'joint denial annihilator']
  ),
  when(
    [isValue(boolean(true)), _],
    [boolean(false), rule`${boolean(false)}`, 'joint denial annihilator']
  ),
  when(
    [_, isValue(boolean(false))],
    (l, _r) => [not(unit(l)), notRule(l), 'joint denial complementation']
  ),
  when(
    [isValue(boolean(false)), _],
    (_l, r) => [not(unit(r)), notRule(r), 'joint denial complementation']
  ),
  when(
    deepEquals, 
    (l, _r) => [not(unit(l)), notRule(l), 'joint denial complementation']
  ),
  when<Complement, Complement>(
    (l, r) => isComplement(l) && isComplement(r),
    (l, r) => [
      and(l.expression, r.expression), 
      andRule(l.expression, r.expression), 
      'De Morgan'
    ]
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(false), rule`${boolean(false)}`, 'joint denial annihilator']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(false), rule`${boolean(false)}`, 'joint denial annihilator']
  )
)
