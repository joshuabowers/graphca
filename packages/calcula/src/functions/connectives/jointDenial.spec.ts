import { unit } from '../../monads/writer'
import { expectToEqualWithSnapshot } from '../../utility/expectations'
import { Clades, Genera, Species } from '../../utility/tree'
import { boolean } from '../../primitives'
import { variable } from '../../variable'
import { not } from './complement'
import { and } from './conjunction'
import { nor, $nor } from './jointDenial'

describe('$nor', () => {
  it('generates a Joint Denial for a pair of TreeNode inputs', () => {
    expect(
      $nor(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.connective, species: Species.nor,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('nor', () => {
  it('returns false when given two true things', () => {
    expectToEqualWithSnapshot(
      nor(boolean(true), boolean(true)),
      boolean(false)
    )
  })

  it('returns false if the left argument is false', () => {
    expectToEqualWithSnapshot(
      nor(boolean(false), boolean(true)),
      boolean(false)
    )
  })

  it('returns false if the right argument is false', () => {
    expectToEqualWithSnapshot(
      nor(boolean(true), boolean(false)),
      boolean(false)
    )
  })

  it('returns true if both arguments are false', () => {
    expectToEqualWithSnapshot(
      nor(boolean(false), boolean(false)),
      boolean(true)
    )
  })

  it('returns the complement of the left operand if the right is false', () => {
    expectToEqualWithSnapshot(
      nor(variable('x'), boolean(false)),
      not(variable('x'))
    )
  })

  it('returns the complement of the right operand if the left is false', () => {
    expectToEqualWithSnapshot(
      nor(boolean(false), variable('x')),
      not(variable('x'))
    )
  })

  it('returns false if the right operand is true', () => {
    expectToEqualWithSnapshot(
      nor(variable('x'), boolean(true)),
      boolean(false)
    )
  })

  it('returns false if the left operand is true', () => {
    expectToEqualWithSnapshot(
      nor(boolean(true), variable('x')),
      boolean(false)
    )
  })

  it('returns the complement of the left operand if left is equivalent to right', () => {
    expectToEqualWithSnapshot(
      nor(variable('x'), variable('x')),
      not(variable('x'))
    )
  })

  it('returns a conjunction of mutually complemented operands', () => {
    expectToEqualWithSnapshot(
      nor(not(variable('x')), not(variable('y'))),
      and(variable('x'), variable('y'))
    )
  })

  it('returns false if the right operand is the complement of the left', () => {
    expectToEqualWithSnapshot(
      nor(variable('x'), not(variable('x'))),
      boolean(false)
    )
  })

  it('returns false if the left operand is the complement of the right', () => {
    expectToEqualWithSnapshot(
      nor(not(variable('x')), variable('x')),
      boolean(false)
    )
  })

  it('returns a JointDenial on variable input', () => {
    expectToEqualWithSnapshot(
      nor(variable('x'), variable('y')),
      $nor(variable('x'), variable('y'))[0]
    )
  })
})
