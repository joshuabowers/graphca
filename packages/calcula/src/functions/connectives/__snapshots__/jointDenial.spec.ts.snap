// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nor returns a JointDenial on variable input 1`] = `
[
  [
    "(x↓y)",
    "identified joint denial",
  ],
  [
    "([x]↓y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x↓{y})",
    "processed right operand",
  ],
  [
    "(x↓y)",
    "created joint denial",
  ],
]
`;

exports[`nor returns a conjunction of mutually complemented operands 1`] = `
[
  [
    "(¬(x)↓¬(y))",
    "identified joint denial",
  ],
  [
    "([¬(x)]↓¬(y))",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}↓[¬(y)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(y)",
    "identified complement",
  ],
  [
    "¬([y])",
    "processing argument",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "¬({y})",
    "processed argument",
  ],
  [
    "¬(y)",
    "created complement",
  ],
  [
    "(¬(x)↓{¬(y)})",
    "processed right operand",
  ],
  [
    "(¬(x)↓¬(y))",
    "De Morgan",
  ],
  [
    "(x∧y)",
    "identified conjunction",
  ],
  [
    "([x]∧y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∧{y})",
    "processed right operand",
  ],
  [
    "(x∧y)",
    "created conjunction",
  ],
]
`;

exports[`nor returns false if the left argument is false 1`] = `
[
  [
    "(false↓true)",
    "identified joint denial",
  ],
  [
    "([false]↓true)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}↓[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(false↓{true})",
    "processed right operand",
  ],
  [
    "(false↓true)",
    "joint deniable annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns false if the left operand is the complement of the right 1`] = `
[
  [
    "(¬(x)↓x)",
    "identified joint denial",
  ],
  [
    "([¬(x)]↓x)",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}↓[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(¬(x)↓{x})",
    "processed right operand",
  ],
  [
    "(¬(x)↓x)",
    "joint deniable contradiction",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns false if the left operand is true 1`] = `
[
  [
    "(true↓x)",
    "identified joint denial",
  ],
  [
    "([true]↓x)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}↓[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(true↓{x})",
    "processed right operand",
  ],
  [
    "(true↓x)",
    "joint deniable annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns false if the right argument is false 1`] = `
[
  [
    "(true↓false)",
    "identified joint denial",
  ],
  [
    "([true]↓false)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}↓[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(true↓{false})",
    "processed right operand",
  ],
  [
    "(true↓false)",
    "joint deniable annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns false if the right operand is the complement of the left 1`] = `
[
  [
    "(x↓¬(x))",
    "identified joint denial",
  ],
  [
    "([x]↓¬(x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[¬(x)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "(x↓{¬(x)})",
    "processed right operand",
  ],
  [
    "(x↓¬(x))",
    "joint deniable contradiction",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns false if the right operand is true 1`] = `
[
  [
    "(x↓true)",
    "identified joint denial",
  ],
  [
    "([x]↓true)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(x↓{true})",
    "processed right operand",
  ],
  [
    "(x↓true)",
    "joint deniable annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns false when given two true things 1`] = `
[
  [
    "(true↓true)",
    "identified joint denial",
  ],
  [
    "([true]↓true)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}↓[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(true↓{true})",
    "processed right operand",
  ],
  [
    "(true↓true)",
    "joint deniable annihilator",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`nor returns the complement of the left operand if left is equivalent to right 1`] = `
[
  [
    "(x↓x)",
    "identified joint denial",
  ],
  [
    "([x]↓x)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(x↓{x})",
    "processed right operand",
  ],
  [
    "(x↓x)",
    "joint deniable complementation",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`nor returns the complement of the left operand if the right is false 1`] = `
[
  [
    "(x↓false)",
    "identified joint denial",
  ],
  [
    "([x]↓false)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}↓[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(x↓{false})",
    "processed right operand",
  ],
  [
    "(x↓false)",
    "joint deniable complementation",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`nor returns the complement of the right operand if the left is false 1`] = `
[
  [
    "(false↓x)",
    "identified joint denial",
  ],
  [
    "([false]↓x)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}↓[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(false↓{x})",
    "processed right operand",
  ],
  [
    "(false↓x)",
    "joint deniable complementation",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
]
`;

exports[`nor returns true if both arguments are false 1`] = `
[
  [
    "(false↓false)",
    "identified joint denial",
  ],
  [
    "([false]↓false)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}↓[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(false↓{false})",
    "processed right operand",
  ],
  [
    "(false↓false)",
    "joint deniable complementation",
  ],
  [
    "¬(false)",
    "identified complement",
  ],
  [
    "¬([false])",
    "processing argument",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "¬({false})",
    "processed argument",
  ],
  [
    "¬(false)",
    "boolean complement",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;
